### PR TITLE
Improve say, add listen command and F5 messages

### LINKED ISSUE #512: Check errors in win condition
If an exception occurs in the win condition, we do not test it:
```Haskell
-- Execute the win condition check *hypothetically*: i.e. in a
-- fresh CESK machine, using a copy of the current game state.
v <- runThrow @Exn . evalState @GameState g $ evalPT t
case v of
  Left _exn -> return () -- TODO: Check errors in win condition
  Right (VBool True) -> winCondition .= Won False
  _ -> return ()
```
Possible solutions:
- keep the robot around and add the error message to its log
- put the error message to the message queue (used by `say`) and check that
- add an option for tests to cause this to throw an IO error

### LINKED ISSUE #513: Revive message queue
From #26:
> I was thinking `say` would become useful as a global God log (we could read `messageQueue` in some pop-up creative window) and check in the tests (no log required). Maybe other robots could hear you too if they are nearby and `listen`.

- add an <kbd>F7</kbd> window where the message queue would be shown in cheating/creative mode
- consider using `LogEntry` instead of raw `Text`, which could help with message formatting:
    ```Haskell
    Say -> case vs of
      [VString msg] -> do
        rn <- use robotName -- TODO: use robot name + ID
        emitMessage (T.concat [rn, ": ", msg])
    ```
- bonus: add a `listen: cmd (robot * string)` command that would allow you to hear other robots `say` something

### LINKED ISSUE #577: Add listen command
Once #565 is merged it would be nice to also allow robots to hear each other and react programmatically.

## Proposed solution
- add the command `listen: cmd string`
  - it will get the first message said by any robot within a radius of 32 cells
  - if there are more messages, take the closest one
  - if there are still more messages, take one at random or jumble them together if you are feeling evil :smiling_imp:
- when a message is obtained, add it to the robot log (and filter message notifications to unique)
  - maybe drop the `Bool` from message notifications and add a field if the message was logged or said to the `LogEntry`?
- create a hearing capability provided by some "listening pipe" (pipe + copper wire + paper?) or something better

## Alternatives
- the command could take an `int` timeout parameter or we could have separate `timeout` command
- the command could fold over messages
- the command could wait for only one tick and have return type `() + string`

### PR DESCRIPTION
**Pull request by xsebek**
- use `LogEntry` for said messages
- add fields to log entries for sorting
- add a messages panel that also shows logs (logger required)
  - colour messages (also in logger)
    - logs bold, message colour per robot and exceptions red
  - show message time
- test notifications for new messages
- add `listen` command and capability
  - add a hearing aid device that provides the capability
- move the F-key modal hints to the top left world corner
- close #513 
- close #512
- close #577

### CODE REVIEW

**Comment by xsebek** [comment]
There is a problem with the panel, it does not scroll past the first page and a little:
```
def for = \i.\c. if (i > 0) {c i; for (i-1) c} {} end
for 10 (\i. log ("index" ++ format i))
for 50 (\i. build {say ("index" ++ format i)})
```
Screenshot from 2022-07-18 20-02-31

**Review by byorgey — APPROVED** [review]
Looks great!  I am not sure what is causing the scroll limit on the messages dialog, I will give that some more thought.

File: src/Swarm/TUI/View.hs [hunk 1]
```diff
@@ -511,6 +514,25 @@ descriptionTitle e = " " ++ from @Text (e ^. entityName) ++ " "
 descriptionWidget :: AppState -> Entity -> Widget Name
 descriptionWidget s e = padLeftRight 1 (explainEntry s e)
 
+messagesWidget :: GameState -> Widget Name
+messagesWidget gs = viewport MessageViewport Vertical (vBox widgetList)
+ where
+  widgetList = focusNewest . map drawLogEntry' . toList . Seq.sort $ focusedLogs <> messages
```
Comment (TristanCacqueray) [inline src/Swarm/TUI/View.hs:520]: This seems to fix the scroll issue, not sure why though :)
```suggestion
  widgetList = map drawLogEntry' . toList . Seq.sort $ focusedLogs <> messages
```

File: src/Swarm/TUI/View.hs [hunk 2]
```diff
@@ -551,11 +573,16 @@ drawKeyMenu s =
...
-      <> notificationKey availableRecipes "F3" "Recipes"
-      <> notificationKey availableCommands "F4" "Commands"
-      <> [(NoHighlight, "^v", "creative") | cheat]
-      <> [(NoHighlight, "^g", "goal") | goal]
+    catMaybes
+      [ Just (NoHighlight, "F1", "help")
+      , Just (NoHighlight, "F2", "robots")
+      , notificationKey availableRecipes "F3" "Recipes"
+      , notificationKey availableCommands "F4" "Commands"
+      , Just (NoHighlight, "F4", "Commands")
```
Comment (TristanCacqueray) [inline src/Swarm/TUI/View.hs:581]: This seems to cause the F4 command help to appear twice.

[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply src/Swarm/TUI/View.hs:581]: Perhaps we should use the same notification logic so that the button is highlighted when there are new messages?

File: src/Swarm/TUI/View.hs [hunk 3]
```diff
@@ -511,6 +514,25 @@ descriptionTitle e = " " ++ from @Text (e ^. entityName) ++ " "
 descriptionWidget :: AppState -> Entity -> Widget Name
 descriptionWidget s e = padLeftRight 1 (explainEntry s e)
 
+messagesWidget :: GameState -> Widget Name
+messagesWidget gs = viewport MessageViewport Vertical (vBox widgetList)
+ where
+  widgetList = focusNewest . map drawLogEntry' . toList . Seq.sort $ focusedLogs <> messages
+  focusNewest = over (element 1) visible
```
Comment (byorgey) [inline src/Swarm/TUI/View.hs:521]: Explicitly setting `visible` on the newest log message is what is causing the scrolling not to work.  `visible` specifically demands that the given element is visible, so you cannot scroll it out of the window.  If I remove the `focusNewest` call then scrolling works as expected.

Comment (byorgey) [reply src/Swarm/TUI/View.hs:521]: Also, isn't `element 1` the *second* newest log message?  That's why you can scroll the window by one line but no more.

Comment (xsebek) [reply src/Swarm/TUI/View.hs:521]: Hmm, we use it in the other log list, how is it unset then?

I wanted to go back to that lens and try `_1` or head or something, never did I expect it would behave so weirdly. 🤦

File: src/Swarm/TUI/View.hs [hunk 4]
```diff
@@ -511,6 +514,25 @@ descriptionTitle e = " " ++ from @Text (e ^. entityName) ++ " "
...
+      hBox
+        [ txt $ "[" <> view leRobotName e <> "] "
+        , txtWrapWith indent2 (e ^. leText)
+        ]
+  focusedLogs = (,True) <$> gs ^. robotMap . ix (gs ^. focusedRobotID) . robotLog
+  messages = (,False) <$> (if gs ^. creativeMode then id else Seq.filter noOther) (gs ^. messageQueue)
+  noOther e = gs ^. focusedRobotID == e ^. leRobotID
+  -- color each robot message with different color of the world (except those with background)
+  robotColor rid = fgCols !! (rid `mod` fgColLen)
+  fgCols = filter (`notElem` [waterAttr, iceAttr]) worldAttributes
```
[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/TUI/View.hs:533]: Yeah, I should filter it properly and check that background is default.

File: src/Swarm/TUI/View.hs [hunk 5]
```diff
@@ -511,6 +514,25 @@ descriptionTitle e = " " ++ from @Text (e ^. entityName) ++ " "
...
+  focusNewest = over (element 1) visible
+  drawLogEntry' (e, isLog) =
+    withAttr (if isLog then notifAttr else robotColor (e ^. leRobotID)) $
+      hBox
+        [ txt $ "[" <> view leRobotName e <> "] "
+        , txtWrapWith indent2 (e ^. leText)
+        ]
+  focusedLogs = (,True) <$> gs ^. robotMap . ix (gs ^. focusedRobotID) . robotLog
+  messages = (,False) <$> (if gs ^. creativeMode then id else Seq.filter noOther) (gs ^. messageQueue)
+  noOther e = gs ^. focusedRobotID == e ^. leRobotID
```
[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/TUI/View.hs:530]: Exactly, though its hard to get right given that robots move.

I have some ideas about how listen or always-on-listen could work, but I will have to think it through.

File: src/Swarm/TUI/View.hs (hunk 1 above)
Comment (TristanCacqueray) [reply src/Swarm/TUI/View.hs:520]: It doesn't like the scroll is working properly yet.

Comment (xsebek) [reply src/Swarm/TUI/View.hs:520]: Fixed :white_check_mark:

File: src/Swarm/TUI/View.hs (hunk 3 above)
Comment (xsebek) [reply src/Swarm/TUI/View.hs]: Fixed :white_check_mark:

**Comment by xsebek** [comment]
@TristanCacqueray  I tried adding a button to allow (un)pausing the game from the message view, but it runs into some problem mentioned in `maybeUnpause`:
```Haskell
  -- When unpausing, it is critical to ensure the next frame doesn't
  -- catch up from the time spent in pause.
  -- TODO: manage unpause more safely to also cover
  -- the world event handler for the KChar 'p'.
```
Should I just make `maybeUnpause` toggle the run status to also set the time in the world event handler and my message queue?

**Comment by xsebek** [comment]
Scrolling now works somewhat intuitively (unpaused autoscroll, paused unobstructed scroll), please give it a try! :slightly_smiling_face: 
Screenshot from 2022-07-19 02-27-04

**Comment by xsebek** [comment]
I squashed and rebased the previous commits.

The last commit should fix the unpausing, please check it out as it's easy to break. I think I got it right on the 3rd try though. :sweat_smile: 

I will try to use the notification logic and then it should be good to merge.

Thanks, @byorgey  and @TristanCacqueray, for helping me with this UI work. :slightly_smiling_face: 

Finally, I can't wait to try out #340, having `AppState` as the monad state just makes sense and I will not miss `continue`.

File: src/Swarm/TUI/Controller.hs [hunk 6]
```diff
@@ -319,6 +319,7 @@ handleModalEvent s = \case
     case s ^? uiState . uiModal . _Just . modalDialog . to dialogSelection of
       Just (Just QuitButton) -> quitGame s'
       Just (Just (NextButton scene)) -> startGame scene s'
+      Just (Just PauseButton) -> continue $ s & gameState . runStatus %~ toggleRunStatus
```
[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/TUI/Controller.hs]: Should work now! :heavy_check_mark:

File: src/Swarm/TUI/Attr.hs [hunk 7]
```diff
@@ -132,6 +132,31 @@ sepAttr = "sep"
 infoAttr = "info"
 defAttr = "def"
 
+worldAttributes :: [AttrName]
```
[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/TUI/Attr.hs:135]: One way to do it is to not allow accessing world attribute names and create a `[(AttrName, Attr)]` list of world attributes.

Those I could filter and pass directly to the attribute map. :smiling_imp:

**Review by byorgey — APPROVED** [review]
This looks good to me, and seems to work properly except for exiting a modal via the `Esc` key (see my other comment).

**Comment by byorgey** [comment]
@xsebek #561 will be great, go ahead and give it a shot!

File: src/Swarm/TUI/Attr.hs (hunk 7 above)
[the remark this answers is not in the document]
Comment (byorgey) [reply src/Swarm/TUI/Attr.hs:135]: Yeah, we really don't need Haskell constants, I think we should get rid of them.  We can easily look up attributes by name from some kind of map or association list.

File: src/Swarm/Game/Step.hs [hunk 8]
```diff
@@ -110,7 +110,7 @@ gameTick = do
 
             when (newLoc /= oldLoc) $ do
               robotsByLocation . at oldLoc %= deleteOne rn
-              robotsByLocation . at newLoc . non Empty %= IS.insert rn
+              robotsByLocation . at newLoc . non Empty %= IS.insert rn -- TODO: should this be above?
```
[the remark this answers is not in the document]
Comment (byorgey) [reply src/Swarm/Game/Step.hs]: I'm not sure what you mean by "should be above", do you mean the `non Empty` should be on the previous line instead, with `oldLoc`?  I don't think that makes sense.  `at` yields a `Maybe Set`; `deleteOne` explicitly takes a `Maybe Set` because we want it to explicitly remove the mapping entirely if the set becomes empty.  On the other hand, the `non Empty` on the second line turns the `Maybe Set` into a `Set`, substituting an empty set for `Nothing`, so that if no mapping existed for the new location previously, it creates an empty set for us to insert `rn` into.

[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/Game/Step.hs]: Oh, I thought it would remove the empty set, so I had it backwards. Thanks for the explanation :slightly_smiling_face:

File: src/Swarm/TUI/Controller.hs [hunk 9]
```diff
@@ -214,7 +214,7 @@ handleMainEvent s = \case
     continue s
   Key V.KEsc
     | isJust (s ^. uiState . uiError) -> continue $ s & uiState . uiError .~ Nothing
-    | isJust (s ^. uiState . uiModal) -> maybeUnpause s >>= (continue . (uiState . uiModal .~ Nothing))
+    | isJust (s ^. uiState . uiModal) -> safeTogglePause s >>= (continue . (uiState . uiModal .~ Nothing))
```
[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/TUI/Controller.hs]: Should be fixed now :+1:

File: src/Swarm/TUI/Controller.hs [hunk 10]
```diff
@@ -223,7 +223,15 @@ handleMainEvent s = \case
...
-  FKey 5 -> toggleModal s MessagesModal >>= continue
+  FKey 5 -> do
+    let t = s ^. gameState . ticks
+    s' <- toggleModal s MessagesModal
+    liftIO $ appendFile "/tmp/debug.log" $ "\n\nOLD LEN - " <> show (s ^. gameState . messageNotifications . notificationsCount)
+    liftIO $ appendFile "/tmp/debug.log" $ "\nOLD TIME - " <> show (s' ^. gameState . lastSeenMessageTime)
+    liftIO $ appendFile "/tmp/debug.log" $ "\nNEW TIME - " <> show t
+    liftIO $ appendFile "/tmp/debug.log" $ "\nNEW LEN - " <> show (s' ^. gameState . messageNotifications . notificationsCount)
+    liftIO $ appendFile "/tmp/debug.log" $ "\nMESSAGES - " <> show (view leTime <$> s' ^. gameState . messageQueue)
+    continue (s' & gameState . lastSeenMessageTime .~ t)
```
Comment (xsebek) [inline src/Swarm/TUI/Controller.hs:234]: Somehow, I am not counting the number of new messages right and it's always zero! :facepalm: 

The flow in Controller should be correct, so it has to be some stupid mistake in State.

Comment (xsebek) [reply src/Swarm/TUI/Controller.hs]: Should work now! :heavy_check_mark:

File: src/Swarm/TUI/View.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/TUI/View.hs]: Done! :white_check_mark:

File: src/Swarm/TUI/View.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/TUI/View.hs]: Let's do it in separate PR.

File: src/Swarm/TUI/Attr.hs (hunk 7 above)
[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/TUI/Attr.hs:135]: Well, I meant keeping constants _only_ for those we use in the Haskell code and _**not** those used in the world map*_.

<sup>*) With the exception of the few attributes that we use for drawing the logo.</sup>

It would not be great to have outdated attribute strings in code, so I think it is a good idea to keep those. :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/TUI/Attr.hs]: @byorgey I just pushed my solution, what do you think? :slightly_smiling_face:

File: src/Swarm/TUI/View.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/TUI/View.hs]: The code should now prevent this, all world attributes have only foreground colour and water is the only exception.

File: src/Swarm/Game/State.hs [hunk 11]
```diff
@@ -440,7 +445,7 @@ viewCenterRule = lens getter setter
     case rule of
       VCLocation v2 -> g {_viewCenterRule = rule, _viewCenter = v2}
       VCRobot rid ->
-        let robotcenter = g ^? robotMap . ix rid <&> view robotLocation
+        let robotcenter = g ^? robotMap . ix rid . robotLocation
          in -- retrieve the loc of the robot if it exists, Nothing otherwise.
             -- sometimes, lenses are amazing...
             case robotcenter of
```
Comment (xsebek) [inline src/Swarm/Game/State.hs:451]: _Sometimes, lenses are amazing..._ and sometimes they give you a headache, trying to find out why is the special robot that you added (later) to the game state not focused. :smile:

File: src/Swarm/TUI/View.hs [hunk 12]
```diff
@@ -84,6 +80,10 @@ import Swarm.TUI.Border
 import Swarm.TUI.Model
 import Swarm.TUI.Panel
 import Swarm.Util
+import System.Clock (TimeSpec (..))
+import Text.Printf
+import Text.Wrap
+import Witch (from)
```
[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/TUI/View.hs:85]: Done in #579! :+1:

**Review by byorgey — COMMENTED** [review]
What is the difference between `log` and `say`?  I'm sure there is a difference, but I don't think I am clear on exactly what the difference is.

File: src/Swarm/Game/State.hs (hunk 11 above)
Comment (xsebek) [reply src/Swarm/Game/State.hs:451]: The lenses I used were ordered with `.` not `&`, so when I added a robot on line X and called this lens on line X+1 it did not do anything.

It did not really matter, because there were other issues, so it was more of a funny discovery 😄

File: src/Swarm/TUI/Controller.hs [hunk 13]
```diff
@@ -223,7 +223,10 @@ handleMainEvent s = \case
   FKey 4 | not (null (s ^. gameState . availableCommands . notificationsContent)) -> do
     s' <- toggleModal s CommandsModal
     continue (s' & gameState . availableCommands . notificationsCount .~ 0)
-  FKey 5 -> toggleModal s MessagesModal >>= continue
+  FKey 5 -> do
+    let t = s ^. gameState . ticks
+    s' <- toggleModal s MessagesModal
+    continue (s' & gameState . lastSeenMessageTime .~ t)
```
[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/TUI/Controller.hs:229]: Oh good point, I will fix that. 👍

File: src/Swarm/Language/Syntax.hs [hunk 14]
```diff
@@ -555,9 +555,12 @@ constInfo c = case c of
     command 0 long . doc "Deconstruct an old robot." $
       ["Salvaging a robot will give you its inventory, installed devices and log."]
   Say ->
-    command 1 short . doc "Emit a message." $ -- TODO: #513
-      [ "The message will be in a global log, which you can not currently view."
-      , "https://github.com/swarm-game/swarm/issues/513"
+    command 1 short . doc "Emit a message." $
+      [ "The messages can be viewed in the message window."
+      , "However, there is a limit to how distant messages you can hear."
```
[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/Language/Syntax.hs:560]: The problem with showing all messages in that place is that it might give the wrong impression that the robot was able to hear them. But it could have just moved to the radius of that message or conversely not see messages it would have been able to hear!

My plan was to be strict and just include those that the robot heard using the `listen` command. So just removing the duplicity in Creative mode and colouring those logs that are saying "I heard robot X say Y".

But that is not very user-friendly and in hindsight, I should show at least some messages.

A truly ~evil~ realistic thing to do would be to show the messages whose distance equals their age. Meaning that sound travels one cell each tick.
Litlle less ~evil~ realistic would be to have some interval like one real-world second to make this more user-friendly.
I am starting to like this idea - it can populate the message queue history and relativity is just cool.
Bonus points for checking that the message has a path to the robot that is not blocked! :rofl: 

Easier to implement would be to just show the messages in the radius and limit them to last real-world second - so 16 ticks.
But that is a bit annoying and does not make much sense.

Whichever way we choose, I feel it should be how `listen` finds messages, so that the players can use the message queue as a guide to how 

---

**TL;DR:** I propose I will implement one of the following options, which will form the logic of `listen`:
- show say messages of other robots in radius _and time limit of one real-world second_
- show say messages that would reach the robot travelling at one cell per tick
- show say messages that would reach the robot travelling at one cell per tick _with a tolerance of one real-world second_
- _show all messages_ as we are not pedantic relativity nerds and #577 is not implemented

What do you think @byorgey?

**Comment by xsebek** [comment]
> @byorgey What is the difference between log and say? I'm sure there is a difference, but I don't think I am clear on exactly what the difference is.

Easy - log writes to the logger, otherwise, it fails. Said messages go to the global message queue and do not require anything.

---

The current logic for showing things in the panel is:
- show focused robots logs and messages - it would know what it logged and said
- in creative mode show all messages (but only focused robot logs)

As you pointed out, that may be too pedantic, see [my ideas](https://github.com/swarm-game/swarm/pull/565#discussion_r928314003) about how to improve the situation.

---

P.S. The reason why I implemented it so was that I just wanted a way to debug scenarios and win conditions without the hassle of inspecting the installed robot's logger. But making it more useful outside creative mode should be doable in this PR.

File: src/Swarm/Language/Syntax.hs (hunk 14 above)
[the remark this answers is not in the document]
Comment (byorgey) [reply src/Swarm/Language/Syntax.hs:560]: Relativity is fun but doesn't seem relevant here, since even audio signals typically travel fast enough that the delay is not really noticeable within the range of the sensor.  When we get to #94 we should definitely come back to that idea.  

How about this: 
- Messages from other robots only show up in the Messages dialog *if you have a listening device installed*
- Then we just show say messages of other robots within the radius and a certain time limit.
    - Am I understanding correctly that by "a time limit of one real-world second" you mean that messages would flash on the Messages dialog for one second and then disappear again?
    - One second seems way too short to see and read any messages.
    - Why one real-world second?  That would be more complex to implement and I'm not sure what the benefit would be.  Why not just say "16 ticks" or whatever?

[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/Language/Syntax.hs:560]: > Am I understanding correctly that by "a time limit of one real-world second" you mean that messages would flash on the Messages dialog for one second and then disappear again?

Exactly, so it's not great. But maybe there is a better option that I forgot about:
- say will put the message to the global queue and to all robot logs of robots that:
  - are within the radius
  - have logger
  - have some listening device

Suggestion for "hearing aid" recipe? :smile: Maybe we can do #577 here too.

[the remark this answers is not in the document]
Comment (byorgey) [reply src/Swarm/Language/Syntax.hs:560]: Ah, right, if we put it in the logger then they will remember it so it doesn't matter if they move later. I like it!

For a hearing aid, maybe something like a copper pipe, some paper, a lodestone, and some wires?

File: src/Swarm/TUI/Controller.hs (hunk 13 above)
[the remark this answers is not in the document]
Comment (xsebek) [reply src/Swarm/TUI/Controller.hs]: Should be fixed now :slightly_smiling_face:

**Comment by xsebek** [comment]
Say and listen should now work:
Screenshot from 2022-08-01 00-30-41

**Comment by xsebek** [comment]
__I think this is ready!__ :smile: 

Say should work pretty well with the new messages window. There are a few tests, but I also tested it manually and that `[F5] Messages` getting highlighted with new messages feels very satisfying. Big thanks to @TristanCacqueray for laying the groundwork for that! :+1: 

Overall the messages window should be easier to read and scroll through than the logger and I hope to improve it with e.g. filters in the future. It reuses a lot of the logic for the other windows and also the time widget. :framed_picture: 

Finally the `listen` command is a bit of new addition and may have some rough edges. But it works well enough in the starting tutorial scenario, so let's hope there are no nasty bugs. :crossed_fingers:

File: data/entities.yaml [hunk 15]
```diff
@@ -816,6 +816,22 @@
   properties: [portable]
   capabilities: [log]
 
+- name: hearing aid
+  display:
+    attr: device
+    char: '@'
+  description:
+  - "Allows robot to hear what nearby robots are saying."
```
Comment (byorgey) [inline data/entities.yaml:824]:
```suggestion
  - "Allows a robot to hear what nearby robots are saying."
```

File: data/entities.yaml [hunk 16]
```diff
@@ -816,6 +816,22 @@
...
+  - "Simply having this device installed will automatically
+     add messages said by nearby robots to this robots log,
```
Comment (byorgey) [inline data/entities.yaml:826]:
```suggestion
     add messages said by nearby robots to this robot's log,
```

File: data/recipes.yaml [hunk 17]
```diff
@@ -478,6 +478,15 @@
   out:
   - [1, strange loop]
 
+- in:
+  - [2, copper pipe]
+  - [4, copper wire]
+  - [1, paper]
+  - [1, lodestone]
+  out:
+  - [2, hearing aid]
```
Comment (byorgey) [inline data/recipes.yaml:487]: Haha, I like that the recipe produces one for each ear :smile:

File: src/Swarm/Game/State.hs [hunk 18]
```diff
@@ -395,40 +458,46 @@ viewCenterRule = lens getter setter
...
+    -- classic players only get to see messages that they said and a one message that they just heard
+    -- other they have to get from log
+    latestMsg = messageIsRecent gs
+    closeMsg = messageIsFromNearby (gs ^. viewCenter)
+    focusedOrLatestClose mq =
+      (Seq.take 1 . Seq.reverse . Seq.filter closeMsg $ Seq.takeWhileR latestMsg mq)
+        <> Seq.filter ((== gs ^. focusedRobotID) . view leRobotID) mq
+
+messageIsRecent :: GameState -> LogEntry -> Bool
+messageIsRecent gs e = e ^. leTime >= gs ^. ticks - 1
```
Comment (byorgey) [inline src/Swarm/Game/State.hs:497]: Does this mean it will only show messages that happened more recently than 1/16 of a second ago?

File: src/Swarm/Game/Step.hs [hunk 19]
```diff
@@ -249,16 +259,28 @@ randomName = do
...
+createLogEntry :: (Has (State GameState) sig m, Has (State Robot) sig m) => Bool -> Text -> m LogEntry
+createLogEntry isSaid msg = do
+  rid <- use robotID
   rn <- use robotName
   time <- use ticks
-  robotLog %= (Seq.|> LogEntry msg rn time)
+  loc <- use robotLocation
+  pure $ LogEntry time isSaid rn rid loc msg
+
+-- | Print some text via the robot's log.
```
Comment (byorgey) [inline src/Swarm/Game/Step.hs:273]: Why do we need the Bool argument? Can we add a comment explaining what it is for?

File: data/entities.yaml [hunk 20]
```diff
@@ -816,6 +816,22 @@
...
+     assuming it has a logger installed."
+  - "That way you can view any heard message later either in
+     the logger or the message window."
+  - "To wait for a message and get the string value, use:"
+  - "`l <- listen; log $ \"I have waited for someone to say \" ++ l`"
+  properties: [portable]
+  capabilities: [log]
```
Comment (byorgey) [inline data/entities.yaml:833]: Should this be `listen`?

File: src/Swarm/TUI/Controller.hs [hunk 21]
```diff
@@ -341,31 +348,45 @@ toggleModal mt = do
...
+  let isRunning = maybe False isRunningModal mt
+  case ev of
+    V.EvKey V.KEnter [] -> do
+      mdialog <- preuse $ uiState . uiModal . _Just . modalDialog
+      toggleModal QuitModal
+      case dialogSelection <$> mdialog of
+        Just (Just QuitButton) -> quitGame
+        Just (Just (NextButton scene)) -> startGame scene
+        _ -> return ()
+    -- TODO: copy pasted here from 'handleMainEvent'
```
Comment (byorgey) [inline src/Swarm/TUI/Controller.hs:375]: What needs to be done here? Per our new policy, we should either do it or turn it into an issue. :smile:

File: src/Swarm/Util.hs [hunk 22]
```diff
@@ -134,6 +142,58 @@ cycleEnum e
...
+--
+-- >>> uniq []
+-- []
+-- >>> uniq [1..5]
+-- [1,2,3,4,5]
+-- >>> uniq (replicate 10 'a')
+-- "a"
+-- >>> uniq "abbbccd"
+-- "abcd"
+uniq :: Eq a => [a] -> [a]
```
Comment (byorgey) [inline src/Swarm/Util.hs:155]: Why do we need a quadratic-time `uniq` rather than an n lg n -time version using `Ord`?

**Review by byorgey — APPROVED** [review]

File: src/Swarm/Util.hs (hunk 22 above)
Comment (xsebek) [reply src/Swarm/Util.hs:155]: This should be linear time unix style uniq, unless I made a mistake. 🙂 

It is used on already sorted list.

File: src/Swarm/TUI/Controller.hs (hunk 21 above)
Comment (xsebek) [reply src/Swarm/TUI/Controller.hs:375]: I would like to avoid duplication, but I am not sure if it is feasible.

The modal handler is plugged in higher than the time handlers, so it overrides them.
But maybe they could be conditionally disabled there and moved up? :thinking:

File: src/Swarm/Game/Step.hs (hunk 19 above)
Comment (xsebek) [reply src/Swarm/Game/Step.hs:273]: Thanks for a reminder, I think a small enumeration `data MessageSource = Said | Logged | ErrorTrace` would work best here and I could colour the errors red. :fire:

File: src/Swarm/Game/State.hs (hunk 18 above)
Comment (xsebek) [reply src/Swarm/Game/State.hs:497]: Oh yes, without a logger it is very weak and only suited for stepping through.

I expect listening robots will likely either have a logger (base, better-equipped relay robots,...) or they will be system robots.

File: data/recipes.yaml (hunk 17 above)
Comment (xsebek) [reply data/recipes.yaml:487]: Thanks, that was the idea. :ear_with_hearing_aid:

File: src/Swarm/Util.hs (hunk 22 above)
Comment (byorgey) [reply src/Swarm/Util.hs:155]: Oh, whoops, that's what I get for reading too quickly. :facepalm:

File: data/entities.yaml (hunk 20 above)
Comment (xsebek) [reply data/entities.yaml:833]: Right... :sweat_smile: 
```suggestion
  capabilities: [listen]
```

File: src/Swarm/Game/Step.hs (hunk 19 above)
Comment (xsebek) [reply src/Swarm/Game/Step.hs:273]: Looks pretty! :art: 
image

**Comment by xsebek** [comment]
@byorgey is it fine if I move the modal help up? The line above REPL was already full.
Screenshot from 2022-08-03 23-08-26

File: src/Swarm/TUI/Controller.hs (hunk 21 above)
Comment (xsebek) [reply src/Swarm/TUI/Controller.hs]: Turns out that was pretty easy and it seems to work well. :+1:

**Comment by byorgey** [comment]
Yeah, I think that's a good idea.

File: src/Swarm/Game/Step.hs [hunk 23]
```diff
@@ -1067,15 +1089,51 @@ execConst c vs s k = do
...
+                | manhattan rl (e ^. leLocation) > manhattan rl (m ^. leLocation) -> es |> m
+                | otherwise -> es |> e
+        let addToRobotLog :: Has (State GameState) sgn m => Robot -> m ()
+            addToRobotLog r = do
+              r' <- execState r $ do
+                hasLog <- hasCapability CLog
+                hasListen <- hasCapability CListen
+                loc' <- use robotLocation
+                when (hasLog && hasListen) (robotLog %= addLatestClosest loc')
+              addRobot r'
```
Comment (kostmo) [inline src/Swarm/Game/Step.hs:1110]: For reference: original definition of `addToRobotLog`
